enhancements/microshift/lvms-topolvm-downstream-replacement.md

```diff
@@ -0,0 +1,248 @@
+an empty deviceClass in the LVMCluster and then coordinating that with the vg-manager.
+We have a prototype implementation that can be pushed into GA in time for the 4.17 release based on
+[OCPEDGE-830](https://issues.redhat.com//browse/OCPEDGE-830) that proves this is possible.
+
+#### Switching from TopoLVM to LVMS
+
+The switch from TopoLVM to LVMS will be done in a single release. We will support
+upgrading from 4.16 to 4.16, and from 4.14 to 4.16 (EUS).
+
+The switch will be done by removing the TopoLVM deployment and installing the LVMS
+deployment. The LVMS deployment will be configured to use the existing deviceClasses
+and storageClasses. There will be leftover images from the TopoLVM deployment that
+can be removed manually from the registry after the upgrade.
+
+There should not be other regressions in the upgrade process, as the LVMS migration
+is transparent to users.
+
+### Risks and Mitigations
+
+There is some risk that the LVMS deployment will not work with the existing
+deviceClasses and storageClasses. We will mitigate this risk by testing the
+upgrade process in CI and by providing a rollback mechanism in case of failure.
+
+Additionally, we will provide documentation on how to correctly set up an empty
+LVMCluster and how to run debugging procedures in case of failure.
+Any issues discovered during this process will be remediated with custom adjustments.
+
+### Drawbacks
+
+The main draw back of this change is that it might require manual remediation
+if a failure occurs during the upgrade process. Since this is a one-time change,
+we believe that the benefits of moving to LVMS outweigh this potential issue.
+
+Additionally, the change will require some additional testing to ensure that
+the upgrade process works as expected. From now on we will need to issue rebase
+updates for the LVMS deployment in MicroShift and that will require adjustments
+
+## Test Plan
+
+We will add an automated test to CI to upgrade the MicroShift version from 4.15 to 4.16 anyhow.
+In this test, we will verify that the upgrade process works as expected and that the LVMS
+deployment is correctly configured to use the existing deviceClasses and storageClasses.
+
+The MicroShift Storage Tests will run regressions and additional tests to ensure that the
+upgrade process works as expected.
+
+Specifically for LVMS, we will introduce a new set of tests that will verify that the
+LVMCluster is correctly configured and that the vg-manager is running as expected.
+
+Additionally, we will provide documentation on how to use the new LVMS deployment so we will
+also test that the documentation is correct and that users can follow it to set up their storage manually.
+
+Last but not least, we will add test cases to the MicroShift Storage Tests that will verify that
+LVMCluster can also be used without ExternalConfiguration and that a regular LVMS deployment would work
+as expected.
```

Comment:
LVMCluster is basically the "hook" at which LVMS starts to initialize volume groups, logical volumes and topolvm controllers. 

Usually LVMCluster can be used to completely abstract the burden of maintaining `lvmd.yaml` because it is created under the hood behind one of its spec fields (one LVMCluster `.spec.deviceClasses[]` entry equals a volumegroup and an entry in `lvmd.yaml`), but with this EP we would allow LVMClusters without any owned deviceClasses, in which case existing deviceClasses from an already present lvmd.yaml would be picked up.

The benefit of LVMCluster is that it gives users the ability to configure VolumeGroups dynamically through a CR and not have to dive into the OS to configure LVM. However, as mentioned in my above comments some highly specialised uses (such as LVM RAID) are not supported in the CR, so thats why users might not want to adopt it.

I discussed this with Daniel before, and we came to the "Hybrid" conclusion that we want to have LVMCluster be there, but ultimately default to not introducing its own volume group management so that the way that MicroShift sets up LVM is not being disturbed.

Reply:
[Here](https://docs.openshift.com/container-platform/4.12/storage/persistent_storage/persistent_storage_local/persistent-storage-using-lvms.html#lvms-creating-lvms-cluster_logical-volume-manager-storage here is a refere) is a useful reference from the docs in case you want to dive a little deeper.

Reply:
These sorts of transitions can be tricky. I think I like the idea of an "inert" LVMCluster API to start. At some point though, I could imagine a user with SNO and MicroShift deployments wanting to use LVMCluster to configure both. It sounds like you're saying that will be possible, _except_ for the [other thread](https://github.com/openshift/enhancements/pull/1601#discussion_r1538226571) where we talked about how today the presence or absence of `lvmd.conf` is what tells MicroShift whether to deploy the CSI driver at all. We should think about the various workflows for cases like:

1. a user is upgrading and has an `lvmd.conf`
2. a user is upgrading and does not have the file (previously indicating CSI is not deployed, what happens now?)
3. a user is deploying on a new host and following our (new?) "best practices" docs

For workflow 3, do we allow them to deploy without CSI? Maybe not? If we do, how does that work?

Reply:
I agree that the transition will likely include a few edge cases. Let me try to answer the workflows you listed briefly:

>     1. a user is upgrading and has an `lvmd.conf`

The Cluster will contain an empty LVMCluster without any deviceClasses. deviceClasses can be added in the CR in which case theyre fully managed, or the traditional solution via lvmd.yaml can be used. Both will work interchangeably and they will not conflict as long as the volume group names do not overlap. If they do, LVMCluster configuration wins.

>     2. a user is upgrading and does not have the file (previously indicating CSI is not deployed, what happens now?)

If the user is upgrading and does not have the file, we did not have the CSI installed before, so there is no reason to install it now. IMO we should leave this cluster alone.

>     3. a user is deploying on a new host and following our (new?) "best practices" docs
> ...
> For workflow 3, do we allow them to deploy without CSI? Maybe not? If we do, how does that work?

My suggestion as per https://github.com/openshift/enhancements/pull/1601/files#diff-e0dc1c4db255868fe6b65d8b8187629884953cb3900650f9b9a755fc24fcd33cR216-R219 would be to not touch the cluster if there is no lvmd.yaml config while we are still using the "embedded" deployment of CSI in Microshift. This means, that for users that are installing a fresh cluster on 4.16, no flow will change. This is because the VG creation/setup is currently handled within microshift and I do not want this to change for the next EL release.

Reply:
> I agree that the transition will likely include a few edge cases. Let me try to answer the workflows you listed briefly:
> 
> > ```
> > 1. a user is upgrading and has an `lvmd.conf`
> > ```
> 
> The Cluster will contain an empty LVMCluster without any deviceClasses. deviceClasses can be added in the CR in which case theyre fully managed, or the traditional solution via lvmd.yaml can be used. Both will work interchangeably and they will not conflict as long as the volume group names do not overlap. If they do, LVMCluster configuration wins.

So after an upgrade the config from lvmd.conf is still applied, and if the user modifies the LVMCluster to add more device classes the 2 sets of device classes are merged?

> > ```
> > 2. a user is upgrading and does not have the file (previously indicating CSI is not deployed, what happens now?)
> > ```
> 
> If the user is upgrading and does not have the file, we did not have the CSI installed before, so there is no reason to install it now. IMO we should leave this cluster alone.
> > ```
> > 3. a user is deploying on a new host and following our (new?) "best practices" docs
> > ```
> > ...
> > For workflow 3, do we allow them to deploy without CSI? Maybe not? If we do, how does that work?
> 
> My suggestion as per https://github.com/openshift/enhancements/pull/1601/files#diff-e0dc1c4db255868fe6b65d8b8187629884953cb3900650f9b9a755fc24fcd33cR216-R219 would be to not touch the cluster if there is no lvmd.yaml config while we are still using the "embedded" deployment of CSI in Microshift. This means, that for users that are installing a fresh cluster on 4.16, no flow will change. This is because the VG creation/setup is currently handled within microshift and I do not want this to change for the next EL release.

OK, so at some point in the future, when we decouple the CSI setup from the main binary and package it separately, instead of looking for `lvmd.conf` to decide if CSI should be deployed we will deploy it if the RPM is installed? We keep coming back to discussions of that decoupling. :-)

In the mean time, if someone wants to use CSI in 4.16 but wants to use the LVMCluster API, they need to create the lvmd.conf file to ensure CSI is deployed? Maybe even an empty file is OK, as a temporary approach?

Reply:
> So after an upgrade the config from lvmd.conf is still applied, and if the user modifies the LVMCluster to add more device classes the 2 sets of device classes are merged?

Yes thats correct, we can handle with with both at the same time, we will merge them together at runtime and update lvmd.yaml as necessary for the ones coming from LVMCluster

Reply:
Since there havent been new questions out of this, do we require some additional docs here or can we close this thread?

Reply:
Closing out as I think this is addressed.